### PR TITLE
fix: COA Batch, Batch Number, & Package Tags in Reports

### PR DESCRIPTION
**Pull request by hitenshKharva**
Ref : [Asana Task](https://app.asana.com/0/1199082161983365/1200218432810657)

Stock Ledger
image

Stock Ageing
image

Batch-Wise Balance
image

### CODE REVIEW

File: erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py [hunk 1]
```diff
@@ -29,7 +30,8 @@ def execute(filters=None):
 							flt(qty_dict.out_qty, float_precision), flt(qty_dict.bal_qty, float_precision),
 							item_map[item]["stock_uom"]
 						])
-
+					if item in package:
```
Comment (rvpasari) [inline erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py]: add comment please

File: erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py [hunk 2]
```diff
@@ -104,3 +106,9 @@ def get_item_details(filters):
 		item_map.setdefault(d.name, d)
 
 	return item_map
+
+def get_package_tag_batch_details(filters):
```
Comment (rvpasari) [inline erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py]: what is the purpose of this function? what values does it take in and return? Please add docstrings

File: erpnext/stock/report/stock_ageing/stock_ageing.py [hunk 3]
```diff
@@ -33,8 +33,7 @@ def execute(filters=None):
 			row.append(details.warehouse)
 
 		row.extend([item_dict.get("total_qty"), average_age,
-			earliest_age, latest_age, details.stock_uom])
-
+			earliest_age, latest_age, details.stock_uom,item_dict.get("package_tag"),details.batch_no,item_dict.get("coa_batch_no")])
```
Comment (rvpasari) [inline erpnext/stock/report/stock_ageing/stock_ageing.py]:
```suggestion
			earliest_age, latest_age, details.stock_uom, item_dict.get("package_tag"), details.batch_no,item_dict.get("coa_batch_no")])
```

File: erpnext/stock/report/stock_ageing/stock_ageing.py [hunk 4]
```diff
@@ -195,7 +216,11 @@ def get_fifo_queue(filters, sle=None):
 			item_details[key]["total_qty"] = d.actual_qty
 		else:
 			item_details[key]["total_qty"] += d.actual_qty
-
+		
+		if key in package:
+			item_details[key]["package_tag"] = package[key]["package_tag"]
+			item_details[key]["coa_batch_no"] = package[key]["coa_batch_no"]
+	print(item_details)
```
Comment (rvpasari) [inline erpnext/stock/report/stock_ageing/stock_ageing.py]:
```suggestion

```

File: erpnext/stock/report/stock_ageing/stock_ageing.py [hunk 5]
```diff
@@ -230,3 +255,9 @@ def get_sle_conditions(filters):
 			where wh.lft >= {0} and rgt <= {1})""".format(lft, rgt))
 
 	return "and {}".format(" and ".join(conditions)) if conditions else ""
+
+def get_package_tag_batch_details(filters):
```
Comment (rvpasari) [inline erpnext/stock/report/stock_ageing/stock_ageing.py]: why are we duplicating this code? Can we not call it from a single place?

File: erpnext/stock/report/stock_balance/stock_balance.py [hunk 6]
```diff
@@ -81,7 +83,12 @@ def execute(filters=None):
 					stock_ageing_data['latest_age'] = date_diff(to_date, fifo_queue[-1][1])
 
 				report_data.update(stock_ageing_data)
-
+			
+			if item in package:
```
Comment (rvpasari) [inline erpnext/stock/report/stock_balance/stock_balance.py]: add comment

File: erpnext/stock/report/stock_balance/stock_balance.py [hunk 7]
```diff
@@ -308,3 +319,9 @@ def get_variant_values_for(items):
 			attribute_map[attr['parent']].update({attr['attribute']: attr['attribute_value']})
 
 	return attribute_map
+
+def get_package_tag_batch_details(filters):
```
Comment (rvpasari) [inline erpnext/stock/report/stock_balance/stock_balance.py]: same, let us not duplicate code

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 8]
```diff
@@ -59,6 +60,11 @@ def execute(filters=None):
 					"finished_qty": finished_qty,
 				})
 
+		if sle.item_code in package:
```
Comment (rvpasari) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: same, add comment to block

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 9]
```diff
@@ -240,3 +250,10 @@ def get_item_group_condition(item_group):
 			item_group_details.rgt)
 
 	return ''
+
+
+def get_package_tag_batch_details(filters):
```
Comment (rvpasari) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: duplicate

File: erpnext/stock/report/stock_ageing/stock_ageing.py (hunk 5 above)
Comment (rvpasari) [reply erpnext/stock/report/stock_ageing/stock_ageing.py]: This could go to the package_tag.py file imo

**Review by rvpasari — CHANGES_REQUESTED** [review]

**Comment by ebdelisle** [comment]
Someone spelled AGING wrong.  Please correct in the title “Stock Ageing” to “Stock Aging”

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 10]
```diff
@@ -240,3 +250,10 @@ def get_item_group_condition(item_group):
...
+	package = {}
+	for d in frappe.db.sql("select item_code, package_tag, batch_no, coa_batch_no from `tabPackage Tag`", as_dict=1):
```
Comment (hrwX) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: Use ORM

File: erpnext/stock/report/stock_balance/stock_balance.py [hunk 11]
```diff
@@ -308,3 +319,9 @@ def get_variant_values_for(items):
...
+	package = {}
+	for d in frappe.db.sql("select item_code, package_tag, batch_no, coa_batch_no from `tabPackage Tag`", as_dict=1):
```
Comment (hrwX) [inline erpnext/stock/report/stock_balance/stock_balance.py]: - Use ORM

File: erpnext/stock/report/stock_ageing/stock_ageing.py [hunk 12]
```diff
@@ -230,3 +255,9 @@ def get_sle_conditions(filters):
...
+	package = {}
+	for d in frappe.db.sql("select item_code, package_tag, batch_no, coa_batch_no from `tabPackage Tag`", as_dict=1):
```
Comment (hrwX) [inline erpnext/stock/report/stock_ageing/stock_ageing.py]: - Use ORM

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py (hunk 2 above)
Comment (sahil28297) [inline erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py]: why are we passing `filters` to this function if we are not using it?

**Review by sahil28297 — CHANGES_REQUESTED** [review]

File: erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py (hunk 2 above)
Comment (hitenshKharva) [reply erpnext/stock/report/batch_wise_balance_history/batch_wise_balance_history.py]: removed it

File: erpnext/stock/report/stock_ledger/stock_ledger.py [hunk 13]
```diff
@@ -240,3 +250,16 @@ def get_item_group_condition(item_group):
 			item_group_details.rgt)
 
 	return ''
+
+
+def get_package_tag_batch_details():
```
Comment (rvpasari) [inline erpnext/stock/report/stock_ledger/stock_ledger.py]: I believe we are moving this out of stock_ledger.py and to a package_tag.py file inside of erpnext

**Review by rvpasari — CHANGES_REQUESTED** [review]

File: erpnext/stock/report/stock_ageing/stock_ageing.py [hunk 14]
```diff
@@ -7,13 +7,15 @@
...
 from erpnext.stock.doctype.serial_no.serial_no import get_serial_nos
+from erpnext.compliance.doctype.package_tag.package_tag import get_package_tag_batch_details
 
 def execute(filters=None):
 
 	columns = get_columns(filters)
 	item_details = get_fifo_queue(filters)
 	to_date = filters["to_date"]
 	_func = lambda x: x[1]
+	package = get_package_tag_batch_details()
```
Comment (hrwX) [inline erpnext/stock/report/stock_ageing/stock_ageing.py]:
```suggestion
	package_tags = get_package_tag_batch_details()
```
- Rename it at multiple places

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/compliance/doctype/package_tag/package_tag.py [hunk 15]
```diff
@@ -101,4 +100,17 @@ def make_package_tag_from_batch(source_name, target_doc=None):
...
+	return target_doc
+
+@frappe.whitelist()
+def get_package_tag_batch_details():
+	""" Function returns package tag, batch and COA Batch No details from tabPackage Tag for items
+
+	Returns:
+		dict: Dictionary of Package Tag details with item code as key.
+	"""	
+	package = {}
```
Comment (rvpasari) [inline erpnext/compliance/doctype/package_tag/package_tag.py]: rename this - should be package_tag_details or something

**Comment by sahil28297** [comment]
> Someone spelled AGING wrong. Please correct in the title “Stock Ageing” to “Stock Aging”

@ebdelisle Ageing is British English and Aging is American English, both are correct, do you want us to change the name to "Aging"?

**Review by rvpasari — CHANGES_REQUESTED** [review]

**Review by hrwX — APPROVED** [review]

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**